File: logwatcher.py
import codecs
import time
import re

from PyQt5.QtCore import Qt, pyqtSignal, pyqtSlot, QObject, QThread

START_LINES = ['Izaro: Ascend with precision.', 'Izaro: The Goddess is watching.', 'Izaro: Justice will prevail.']
FINISH_LINES = ['Izaro: I die for the Empire!', 'Izaro: Delight in your gilded dungeon, ascendant.', 'Izaro: Your destination is more dangerous than the journey, ascendant.', 'Izaro: Triumphant at last!', 'Izaro: You are free!', 'Izaro: The trap of tyranny is inescapable.']
SECTION_COMPLETE_LINES = ['Izaro: By the Goddess! What ambition!', 'Izaro: Such resilience!', 'Izaro: You are inexhaustible!', 'Izaro: You were born for this!']
ZONE_PREFIX = ['Estate', 'Domain', 'Basilica', 'Mansion', 'Sepulchre', 'Sanitorium']
ZONE_SUFFIX = ['Walkways', 'Path', 'Crossing', 'Annex', 'Halls', 'Passage', 'Enclosure', 'Atrium']
DIRECTION = {'N': [0, -1], 'NE': [0.71, -0.71], 'E': [1, 0], 'SE': [0.71, 0.71], 'S': [0, 1], 'SW': [-0.71, 0.71], 'W': [-1, 0], 'NW': [-0.71, -0.71]}
ZONE_REGEX = re.compile(r'^: You have entered (.*?)\.$')
LOG_REGEX = re.compile(r'^\d+/\d+/\d+ \d+:\d+:\d+.*?\[.*?\] (.*)$')

class LogWatcher(QObject):
  labStart = pyqtSignal()
  labFinish = pyqtSignal()
  labExit = pyqtSignal()
  sectionComplete = pyqtSignal()
  zoneChange = pyqtSignal(str)

  def __init__(self):
    super().__init__()
    self.thread = QThread()
    self.moveToThread(self.thread)
    self.thread.start()

  @pyqtSlot()
  def start(self):
    with codecs.open('C:/Program Files (x86)/Steam/steamapps/common/Path of Exile/logs/Client.txt', 'r', encoding='utf-8') as log:
      while log.readline():
        pass
      print('log watcher ready')
      while True:
        where = log.tell()
        line = log.readline()
        if not line:
          time.sleep(1)
          log.seek(where)
        else:
          m = re.match(LOG_REGEX, line)
          if m:
            log_content = m.group(1).strip()
            m1 = re.match(ZONE_REGEX, log_content)
            if m1:
              zone = m1.group(1)
              affixes = zone.split(' ')
              if (affixes[0] in ZONE_PREFIX and affixes[1] in ZONE_SUFFIX) or zone == 'Aspirant\'s Trial':
                self.zoneChange.emit(m1.group(1))
              else:
                self.labExit.emit()
            elif log_content in START_LINES:
              self.labStart.emit()
            elif log_content in FINISH_LINES:
              self.labFinish.emit()
            elif log_content in SECTION_COMPLETE_LINES:
              self.sectionComplete.emit()

File: common/src/worker/logwatcher.cpp
#include "logwatcher.h"

#ifdef Q_OS_WIN
#include "windows.h"
#include "psapi.h"
#endif

static const QStringList START_LINES {
  "Izaro: Ascend with precision.",
  "Izaro: The Goddess is watching.",
  "Izaro: Justice will prevail.",
};
static const QStringList FINISH_LINES {
  "Izaro: I die for the Empire!",
  "Izaro: Delight in your gilded dungeon, ascendant.",
  "Izaro: Your destination is more dangerous than the journey, ascendant.",
  "Izaro: Triumphant at last!",
  "Izaro: You are free!",
  "Izaro: The trap of tyranny is inescapable.",
};
static const QStringList SECTION_FINISH_LINES {
  "Izaro: By the Goddess! What ambition!",
  "Izaro: Such resilience!",
  "Izaro: You are inexhaustible!",
  "Izaro: You were born for this!",
};
static const QStringList PORTAL_SPAWN_LINES {
  ": A portal to Izaro appears."
};
static const QStringList LAB_ROOM_PREFIX {"Estate", "Domain", "Basilica", "Mansion", "Sepulchre", "Sanitorium"};
static const QStringList LAB_ROOM_SUFFIX {"Walkways", "Path", "Crossing", "Annex", "Halls", "Passage", "Enclosure", "Atrium"};
static const QRegularExpression LOG_REGEX {"^\\d+/\\d+/\\d+ \\d+:\\d+:\\d+.*?\\[.*?(\\d+)\\] (.*)$"};
static const QRegularExpression ROOM_CHANGE_REGEX {"^: You have entered (.*?)\\.$"};

LogWatcher::LogWatcher(ApplicationModel* model)
{
  this->model = model;
  clientPath = model->get_settings()->get_poeClientPath();
  file.reset(new QFile(QDir(clientPath).filePath("logs/Client.txt")));

  timer.setInterval(1000);
  timer.setSingleShot(false);
  timer.start();
  connect(&timer, &QTimer::timeout,
          this, &LogWatcher::work);
}

void LogWatcher::work()
{
  // reset file if client path settings have changed
  auto newClientPath = model->get_settings()->get_poeClientPath();
  if (clientPath != newClientPath) {
    clientPath = newClientPath;
    file.reset(new QFile(QDir(clientPath).filePath("logs/Client.txt")));
  }

  // attempt to open file
  if (!file->isOpen()) {
    if (!file->open(QIODevice::ReadOnly)) {

      // try to detect client
      clientPath = findGameClientPath();
      if (clientPath.isEmpty()) {
        model->update_logFileOpen(false);
        return;
      }
      file.reset(new QFile(QDir(clientPath).filePath("logs/Client.txt")));
      if (!file->open(QIODevice::ReadOnly)) {
        model->update_logFileOpen(false);
        return;
      }
      model->get_settings()->set_poeClientPath(clientPath);
    }
    model->update_logFileOpen(true);
    file->seek(file->size());
  }

  while (true) {
    auto line = file->readLine();
    if (line.isEmpty())
      break;
    parseLine(QString::fromUtf8(line).trimmed());
  }
}

void LogWatcher::parseLine(const QString line)
{
  auto logMatch = LOG_REGEX.match(line);
  if (logMatch.hasMatch()) {
    auto clientId = logMatch.captured(1);

    auto logContent = logMatch.captured(2).trimmed();
    auto roomChangeMatch = ROOM_CHANGE_REGEX.match(logContent);

    if (START_LINES.contains(logContent)) {
      setActiveClient(clientId);
      emit labStarted();

    } else if (roomChangeMatch.hasMatch()) {
      auto roomName = roomChangeMatch.captured(1);
      auto affixes = roomName.split(' ');

      if (roomName == "Aspirants\' Plaza") {
        setActiveClient(clientId);
        emit plazaEntered();

      } else if (roomName == "Aspirant\'s Trial" ||
          (affixes.size() == 2 && LAB_ROOM_PREFIX.contains(affixes[0]) && LAB_ROOM_SUFFIX.contains(affixes[1]))) {
        if (isLogFromValidClient(clientId))
          emit roomChanged(roomName);

      } else {
        if (isLogFromValidClient(clientId))
          emit labExit();
      }

    } else if (FINISH_LINES.contains(logContent)) {
      if (isLogFromValidClient(clientId)) {
        emit sectionFinished();
        emit labFinished();
      }

    } else if (SECTION_FINISH_LINES.contains(logContent)) {
      if (isLogFromValidClient(clientId))
        emit sectionFinished();

    } else if (PORTAL_SPAWN_LINES.contains(logContent)) {
      if (isLogFromValidClient(clientId))
        emit portalSpawned();
    }
  }
}

QString LogWatcher::findGameClientPath()
{
#ifdef Q_OS_WIN
  auto hwnd = FindWindowA("POEWindowClass", nullptr);
  if (!hwnd)
    return QString();

  DWORD pid;
  GetWindowThreadProcessId(hwnd, &pid);

  auto handle = OpenProcess(PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid);
  if (!handle)
    return QString();

  char buf[1024];
  auto r = GetModuleFileNameExA(handle, NULL, buf, 1024);
  QString path = r ? QFileInfo(QString(buf)).dir().absolutePath() : QString();

  CloseHandle(handle);
  return path;
#else
  return QString();
#endif
}

void LogWatcher::setActiveClient(const QString& clientId)
{
  activeClientId = clientId;
}

bool LogWatcher::isLogFromValidClient(const QString& clientId) const
{
  return !model->get_settings()->get_multiclientSupport() || clientId == activeClientId;
}
